core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -559,6 +560,30 @@ class KafkaApis(val requestChannel: RequestChannel,
     }
   }
 
+  case class LeaderNode(leaderId: Int, leaderEpoch: Int, node: Node)
+
+  private def getCurrentLeader(tp: TopicPartition): LeaderNode = {
+    val partitionInfoOrError = replicaManager.getPartitionOrError(tp)
+    var leaderId = -1
+    var leaderEpoch = -1
+    partitionInfoOrError match {
+      case Right(x) =>
+          leaderId = x.leaderReplicaIdOpt.getOrElse(-1)
+          leaderEpoch = x.getLeaderEpoch
+      case Left(x) =>
+        debug(s"Unable to retrieve local leaderId and Epoch with error $x, falling back to metadata cache")
+        val partitionInfo = metadataCache.getPartitionInfo(tp.topic, tp.partition)
+        partitionInfo.foreach { info =>
```

Comment:
This `foreach` loop is overwriting the `leaderId` and `leaderEpoch` each time. Is that intentional? Is there a benefit to looping vs. just grabbing the last entry in the collection?

Reply:
Looking at other uses of partitionInfo I think this is a style choice. There can only be 1 partitionInfo in the getPartitionInfo object, so the forEach should only ever access 1 entry, I think this is just a more succinct way of accessing it.

Reply:
getPartitionInfo returns an option. If it exists, foreach will access it. If it doesn't foreach does nothing. This is a common pattern in scala. Are we considering the case when the partition is not present?

Reply:
We also don't need to set vars here. We could have a statement where we return a tuple.

Reply:
```
partitionInfo =  partitionInfoOrError match {
      case Right(partitionInfo) =>
         partitionInfo
      case Left(error) =>
        debug(s"Unable to retrieve local leaderId and Epoch with error $error, falling back to metadata cache")
        metadataCache.getPartitionInfo(tp.topic, tp.partition) match {
          case Some(partitionInfo) => partitionInfo
          case None => handle case where we don't have the partition
        }
 }
```

Reply:
I think the confusion was that foreach implies a list of elements, but there can only be 1 here. I like the use of tuple and match/case here though, I will update to that

Reply:
Yeah, using foreach for scala options is a common pattern.